src/shims/fs.rs

```diff
@@ -160,6 +161,82 @@ trait EvalContextExtPrivate<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, '
         this.set_last_error(ebadf)?;
         Ok((-1).into())
     }
+
+    fn file_type_to_d_type(&mut self, file_type: std::io::Result<FileType>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+        match file_type {
+            Ok(file_type) => {
+                if file_type.is_dir() {
+                    Ok(this.eval_libc("DT_DIR")?.to_u8()? as i32)
+                } else if file_type.is_file() {
+                    Ok(this.eval_libc("DT_REG")?.to_u8()? as i32)
+                } else if file_type.is_symlink() {
+                    Ok(this.eval_libc("DT_LNK")?.to_u8()? as i32)
+                } else {
+                    // Certain file types are only supported when the host is a Unix system.
+                    // (i.e. devices and sockets) If it is, check those cases, if not, fall back to
+                    // DT_UNKNOWN sooner.
+
+                    #[cfg(unix)]
+                    {
+                        use std::os::unix::fs::FileTypeExt;
+                        if file_type.is_block_device() {
+                            Ok(this.eval_libc("DT_BLK")?.to_u8()? as i32)
+                        } else if file_type.is_char_device() {
+                            Ok(this.eval_libc("DT_CHR")?.to_u8()? as i32)
+                        } else if file_type.is_fifo() {
+                            Ok(this.eval_libc("DT_FIFO")?.to_u8()? as i32)
+                        } else if file_type.is_socket() {
+                            Ok(this.eval_libc("DT_SOCK")?.to_u8()? as i32)
+                        } else {
+                            Ok(this.eval_libc("DT_UNKNOWN")?.to_u8()? as i32)
+                        }
+                    }
+                    #[cfg(not(unix))]
+                    Ok(this.eval_libc("DT_UNKNOWN")?.to_u8()? as i32)
+                }
+            }
+            Err(e) => return match e.raw_os_error() {
+                Some(error) => Ok(error),
```

Comment:
This sets the file type to an error code... that makes no sense at all, right?

Reply:
Most of the time we set the error using `set_last_error` and return a sigil value like `-1`. Not sure if that would also apply here